ral-facilities/datagateway-api
# Allow related entities on DB distinct filter

This PR will close #223 

## Description
This branch allows related entities to be used on the distinct filter on the DB backend. The tests I've added in `test_get_with_filters.py` show the additional functionality this branch brings.

I've moved a lot of the functionality that was already present in `filters.py` into another class, `DatabaseFilterUtilities` - I've done this so both the where filter and distinct filter can make use of the same code. (1b7f1325c433cf0917bdbc8e557468a1f110296e contains the bulk of the class creation, 11f75a20a2dba6aac446fbaeb8d61a3f8a436233 removes most of the pre-existing, non-generic code). I've added a test class to test this class (bc9635153bae681e89962fb37d71c14cf16056e7).

I've separated out functionality of `add_query_join()`, more info in the commit message of a2538ccba6810fbd23c6f2c9d1d09ea89ba8498b

This morning, I properly fixed a bug I attempted to fix on Friday (proper fix at 9021fc865872586441fe710b29f7cfbd9ca87522). I tried to use a flag to limit when `add_query_join()` was called, but that didn't work in cases where there was related distinct fields and a where filter in the same request. f16cb15a4360e6fb818db8409508d6f9352127e0 shows actual cases that this bug fixes.


## Testing Instructions
I've added test cases to test/demonstrate the functionality added here. The main thing is that the functionality added here matches what's available in the ICAT backend so construct some GET requests with distinct fields (mixture of related and unrelated) and check they give the same results. You might find that some entity names need to be referenced differently in DB backend (e.g. `investigationType` which should be `type` according to the ICAT schema) but I'm aiming to fix those issues in #227.

- [ ] Check the distinct functionality is the same in both backends

## Agile Board Tracking
Connect to #223
